File: 2015/10_Elves_Look_Elves_Say/Day10.clj
; Day 10: Elves Look, Elves Say
; https://adventofcode.com/2015/day/10

(ns day10
  (:require [clojure.string :as str]))

(defn look-and-say
  "Generate look-and-say sequence for given input"
  [sequence]
  (loop [result (StringBuilder.)
         remaining (seq sequence)
         current (first remaining)
         count 1]
    (if-let [next (first (rest remaining))]
      (if (= next current)
        (recur result (rest remaining) current (inc count))
        (recur (.append (.append result count) current) 
               (rest remaining) 
               next 
               1))
      (str (.append (.append result count) current)))))

(defn generate-look-and-say-sequence [input iterations]
  {:pre [(re-matches #"^\d+$" input)]}
  (loop [sequence input
         i 0]
    (if (= i iterations)
      (count sequence)
      (recur (look-and-say sequence) (inc i)))))

(defn part1 [input]
  (generate-look-and-say-sequence input 40))

(defn part2 [input]
  (generate-look-and-say-sequence input 50))

(def input "1")
(def input-test "1113222113")

(println "Part 1:" (part1 input))
(println "Part 1:" (part1 input-test))
(println "Part 2:" (part2 input))
(println "Part 2:" (part2 input-test))

File: 2015/10_Elves_Look_Elves_Say/Day10.js
// Day 10: Elves Look, Elves Say
// https://adventofcode.com/2015/day/10

const input = '1';
const inputTest = '1113222113';

function lookAndSay(sequence) {
  let result = '';
  let currentChar = sequence[0];
  let count = 1;

  for (let i = 1; i <= sequence.length; i++) {
    if (i === sequence.length || sequence[i] !== currentChar) {
      result += count + currentChar;
      currentChar = sequence[i];
      count = 1;
    } else {
      count++;
    }
  }

  return result;
}

function generateLookAndSaySequence(input, iterations) {
  let sequence = input;
  for (let i = 0; i < iterations; i++) {
    sequence = lookAndSay(sequence);
  }

  return sequence.length;
}

function part1(input) {
  return generateLookAndSaySequence(input, 40);
}

function part2(input) {
  return generateLookAndSaySequence(input, 50);
}

console.log('Part 1:', part1(input)); // 82350
console.time('Part 1');
console.log('Part 1:', part1(inputTest)); // 252594
console.timeEnd('Part 1');

console.log('Part 2:', part2(input)); // 1166642
console.time('Part 2');
console.log('Part 2:', part2(inputTest)); // 3579328
console.timeEnd('Part 2');
